library/src/main/scala/scala/scalajs/macroImpls/UseAsMacros.scala

```diff
@@ -0,0 +1,304 @@
+
+      def isValidMember(mem: Symbol) = (
+          !JSObjectAncestors(mem.owner) &&
+          !mem.isConstructor &&
+          mem.isMethod &&
+          !mem.asTerm.isParamWithDefault
+      )
+
+      for (mem <- tpe.members if isValidMember(mem)) yield {
+        val (name, explicit) = jsName(mem.asMethod)
+
+        // apply is a special case, since when it is not explicitly specified,
+        // it will result in an actual function application in JavaScript
+        // Scala defined types cannot support this --> fail
+        if (name == "apply" && !explicit) {
+          c.abort(c.enclosingPosition, s"$tpe defines an apply method. This " +
+            "cannot be implemented by any Scala exported type, since it " +
+            "would need to chain Function's prototype.")
+        }
+
+        (name, jsMemberFor(tpe, mem.asMethod))
+      }
+    }
+
+    /** Returns the JavaScript name of a method and whether it is specified
+     *  explicitly or not.
+     */
+    private def jsName(sym: MethodSymbol): (String, Boolean) = {
+      val annot = memberAnnotations(sym).find(annotIs(_, JSNameAnnotation))
+      val name = annot.flatMap(annotStringArg).getOrElse(defaultName(sym))
+      (name, annot.isDefined)
+    }
+
+    /** Returns all exported members of a type */
+    private def exportedMembers(tpe: Type): Map[String, List[JSMember]] = {
+      val exports = tpe.baseClasses.flatMap(exportedDecls(tpe, _))
+
+      // Group exports by name
+      val lists = for {
+        (name, elems) <- exports.groupBy(_._1)
+      } yield {
+        (name, elems.map(_._2).toList)
+      }
+
+      lists.toMap
+    }
+
+    /** All exported declarations of a class.
+     *  (both @JSExportAll and @JSExport)
+     */
+    private def exportedDecls(origTpe: Type, sym: Symbol) = {
+      require(sym.isClass)
+
+      val exportAll = sym.annotations.exists(annotIs(_, JSExportAllAnnotation))
+
+      for {
+        mem <- sym.info.decls if mem.isMethod && !mem.isConstructor
+        name <- exportNames(mem.asMethod, exportAll)
+      } yield {
+        (name, jsMemberFor(origTpe, mem.asMethod))
+      }
+    }
+
+    /** Get the JS member for a method in [[origTpe]] */
+    private def jsMemberFor(origTpe: Type, sym: MethodSymbol): JSMember = {
+      sym.info.asSeenFrom(origTpe, sym.owner) match {
+        case MethodType(List(param), returnType)
+          if returnType.typeSymbol == definitions.UnitClass &&
+             sym.name.decodedName.toString.endsWith("_=") =>
+          JSSetter(param.info)
+
+        case NullaryMethodType(returnType) =>
+          JSGetter(returnType)
+
+        case info: MethodType =>
+          @tailrec
+          def flatParams(tpe: Type, acc: List[JSMethodParam]): JSMethod = {
+            tpe match {
+              case MethodType(params, returnTpe) =>
+                val ps = params map { p =>
+                  JSMethodParam(p.info, p.asTerm.isParamWithDefault)
```

Comment:
They are of a special type (the repeated type). Macros currently do not give you any ability to query for such a type, but the comparisons work properly.